enhancements/machine-api/azure-ultra-disks.md

```diff
@@ -0,0 +1,243 @@
+This will allow users to set Ultra Disks as Data Disks for the Machine. This feature has been made available upstream in Cluster API Provider for Azure (CAPZ) by: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1478
+1. Required configuration for specifying the ability in `AzureMachineProviderSpec` to allow the Azure CSI driver to provision Ultra Disks as PVs for the Machine. See the issue tracking the implementation of this feature in the upstream Cluster API Provider for Azure (CAPZ): https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1852
+
+### User Stories
+
+**Story 1**
+As a developer, I'd like to have the ability to use Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 2**
+As a developer, I'd like to specify some basic specs (Disk Size, Logical Unit Number (LUN), Caching Type) for creating Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 3**
+As a developer, I'd like to have the ability to create Persistent Volume Claims (PVCs) which can dynamically bind to a StorageClass backed by Azure Ultra Disk and to mount them successfully to Pods for my workloads needs.
+
+### API Extensions
+
+#### Extension for Ultra Disks as Data Disks
+
+The API design is taken from the upstream Cluster API Provider for Azure (CAPZ), [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/azuremachine_types.go#L77-L79) and [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/types.go#L465-L483).
+The upstream design is followed as closely as possible to ensure smoother future migrations between MAPI and CAPI.
+
+A new field will be added to the `AzureMachineProviderSpec` struct `DataDisks`. The interface will be a slice of `DataDisk` structs allowing multiple Data Disks and various fields specific to them to be further specified.
+
+A `DataDisk` has one Composite type field, `ManagedDisk` of type `*ManagedDiskParameters`, which is a type already defined in the Machine API spec, used for the `ManagedDisk` field in the `OSDisk` type struct, another `AzureMachineProviderSpec` field.
+
+
+```go
+type AzureMachineProviderSpec struct {
+  // Existing fields will not be modified
+  ...
+
+  // DataDisk specifies the parameters that are used to add one or more data disks to the machine
+  // +optional
+  DataDisks []DataDisk `json:"dataDisks,omitempty"`
+}
+
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+type DataDisk struct {
+  // NameSuffix is the suffix to be appended to the machine name to generate the disk name.
+  // Each disk name will be in format <machineName>_<nameSuffix>.
+  NameSuffix string `json:"nameSuffix"`
+  // DiskSizeGB is the size in GB to assign to the data disk.
+  DiskSizeGB int32 `json:"diskSizeGB"`
```

Comment:
Are there any limits on the size that the disk can be? We could set a min/max value using kubebuilder tags as appropriate if there are

Reply:
By consulting the [docs](https://docs.microsoft.com/en-us/azure/virtual-machines/disks-types#ultra-disk-size):

_"[...] Azure ultra disks offer up to 32 TiB per region per subscription by default, but ultra disks support higher capacity by request. To request an increase in capacity, request a quota increase or contact Azure Support."_

but then also:

_"Ultra disks come in several fixed sizes, ranging from 4 GiB up to 64 TiB."_

So we could set it to `min: 4 GiB`, `max: 64 TiB`?

Reply:
Looks like in upstream they limit it to `32TiB` while doing validation (rather than using Kubebuilder tags): https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2349883d1effe6344690459b40e5dc26138010df/api/v1beta1/azuremachine_validation.go#L113-L116
But if the quota can be increased on request to a maximum of `64 TiB` I'm not sure it's a good idea to set it to `32 TiB`.

Reply:
That range might be too big for a validation

Reply:
It is the representation of `64 TiB` in GB which is `65534`, that should fit in the validation, I found higher numbers across GitHub (e.g. https://github.com/Azure/azure-service-operator/blob/90ce85b99debff37c387b85765a6a553100ee155/api/v1alpha1/eventhub_types.go#L72)

Reply:
Tried:
```
// +kubebuilder:validation:Minimum=4
// +kubebuilder:validation:Maximum=65536
```
and it looks like they work.

Reply:
My suspicion is that we should add a maximum, that feels like it could bite us later down the line if they add bigger storage for specific customers or something. I would go with the minimum only

Reply:
👍